### ISSUE TITLE
Spacing between list elements only works for top sublist element

### ISSUE THREAD

**Issue by DTAG-lnu** [typst/typst#5510]
### Description

I am working with Touying to make lecture slides, but I believe this to be a general problem with lists. I have posted the problem in the forum and user miles-1 also agrees that this might be a problem in Typst. The problem is that when I set the spacing for lists with the following code:

`#set list(
  spacing: 1.2em,
)`

It only applies the spacing, as it seems, to the first subitem. The rest of them are spaced out with a larger (default) spacing. This can be seen in this image:
![Skärmbild från 2024-11-25 15-21-48](https://github.com/user-attachments/assets/231ac116-f4c1-404f-9b52-106be2c5cb02)

I have set up an example in the reproduction URL where I also, with the help of user miles-1, done an example at the end where the position of the spacing is shown to be put differently by  using the code below:

`#show list.item: it => {
  show [ ].func(): [SPACE IN LIST HERE]
  it
}`

He, and I think so too, believe that this shows the unexpected behaviour as the text "SPACE IN LIST HERE" is clearly in different places.

I hope this is something that someone can continue to work on or explain if it is the intended behaviour.

### Reproduction URL

https://typst.app/project/rxAaN1fm8jJuAmo80Adcwm

### Operating system

Web app, Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by laurmaedje** [comment]
If there is no blank line between text and a list item of a tight list, it attaches to the paragraph with normal line spacing (aka leading). The relevant code is here: https://github.com/typst/typst/blob/4f3ba7f8caecc49ae1051f0a0c217a44d3fda876/crates/typst-library/src/model/list.rs#L161-L166

Maybe it should attach with list spacing instead. (The default spacing for tights lists is leading anyway.)

**Comment by DTAG-lnu** [comment]
Yes, I believe you are right. Though I am not in a position to be able to offer any additional code for doing that, so I will have to see if anyone picks it up.

**Comment by xkevio** [comment]
> If there is no blank line between text and a list item of a tight list, it attaches to the paragraph with normal line spacing (aka leading). The relevant code is here:
> 
> https://github.com/typst/typst/blob/4f3ba7f8caecc49ae1051f0a0c217a44d3fda876/crates/typst-library/src/model/list.rs#L161-L166
> 
> Maybe it should attach with list spacing instead. (The default spacing for tight lists is leading anyway.)

I don't think it only affects tight lists, it "correctly" uses paragraph spacing here for a non-tight list, even though list `spacing` is explicitly set.

```typ
#set par(spacing: 0.65em)
#set list(spacing: 1em)

- Linked Lists

  - Classes

  - Methods

  - Different Variants
```
![image](https://github.com/user-attachments/assets/9e6d17a8-9ed2-4cc2-acd1-a92f63aeebb5)

**Comment by laurmaedje** [comment]
This is closely related but slightly different. Here, the block above & below spacing of lists is not affected by list.spacing.

**Comment by xkevio** [comment]
> This is closely related but slightly different. Here, the block above & below spacing of lists is not affected by list.spacing.

Ah I see, similarly, this shouldn't be too hard to fix. Seems like just an oversight from all the block layouting reworks and such though I don't exactly know where that would be.

**Comment by saecki** [comment]
> This is closely related but slightly different. Here, the block above & below spacing of lists is not affected by list.spacing.

That's actually not true, as can be seen in [this example](https://typst.app/project/rti49gOyxfgycoFghLqp1H), the spacing above nested lists (green lines) uses paragraph (leading|spacing), but below the list (red lines) spacing is used:

![Image](https://github.com/user-attachments/assets/3e642720-394d-46ee-8a35-28ac32f02cda)

I think the expected result would be to use list spacing if available, and then fall back to using paragraph (leading|spacing).
Or should paragraph spacing

**Comment by laurmaedje** [comment]
> That's actually not true, as can be seen in [this example](https://typst.app/project/rti49gOyxfgycoFghLqp1H), the spacing above nested lists (green lines) uses paragraph (leading|spacing), but below (red lines) the list spacing is used:

I think where the red lines are, you're observing list spacing from the outer list. See this example:
```typ
#set par(leading: 0em)
#set list(spacing: 4em)
#show list: set block(above: 1em, below: 2em)

- Linked lists
  - Classes // 0em attach leading above here because it's a tight list
  - Methods
  - Different variants // 4em list spacing below this item
                       // -- list spacing from the outer list
  
- Testing 

  - Unit test // 1em list above spacing here because it's not tight
  
  - Setting up
  
  - Examples // only 2em spacing below here
             // --- the list below's below spacing

  No big gap. 
```

<img width="222" alt="Image" src="https://github.com/user-attachments/assets/01324343-2263-40b2-b77f-5b223cd6ac08" />

The list above/spacing applies to both tight and non-tight list, but the [attach spacing generated for tight lists](https://github.com/typst/typst/blob/4f3ba7f8caecc49ae1051f0a0c217a44d3fda876/crates/typst-library/src/model/list.rs#L161-L166) overrides it.

**Comment by laurmaedje** [comment]
To summarize the current state, just to clarify things in my head:
- List spacing is between items
- Above/below is around the list _except_ above tight lists, where we always have leading

The issue here is that list spacing differs from leading and hence the list looks untidy. Wide lists also exhibit this behaviour, but it's easy to fix by also setting above/below spacing. Meanwhile, tight lists will always use leading. This is undesirable in some scenarios, but in principle makes sense if you view the tight list as becoming part of the paragraph (see also https://github.com/typst/typst/issues/3206). If we don't want to use leading, the question is whether list spacing or above spacing is more apprioriate. We could also potentially adjust the above/below spacing of tight lists with a built-in show set rule. 

I'm not sure right now what the correct design is. I would need to think more about this.
